system/lib/libc/dynlink.c

```diff
@@ -327,6 +327,30 @@ void emscripten_dlopen(const char* filename, int flags, void* user_data,
   _emscripten_dlopen_js(p, dlopen_js_onsuccess, dlopen_js_onerror, d);
 }
 
+static void promise_onsuccess(void* user_data, void* handle) {
+  em_promise_t p = (em_promise_t)user_data;
+  dbg("promise_onsuccess: %p", p);
+  emscripten_promise_resolve(p, EM_PROMISE_FULFILL, handle);
+  emscripten_promise_destroy(p);
+}
+
+static void promise_onerror(void* user_data) {
+  em_promise_t p = (em_promise_t)user_data;
+  dbg("promise_onerror: %p", p);
+  emscripten_promise_resolve(p, EM_PROMISE_REJECT, NULL);
+  emscripten_promise_destroy(p);
+}
+
+// emscripten_dlopen_promise is currently implemented on top of the callback
+// based API (emscripten_dlopen).
+// TODO(sbc): Consider inverting this and perhaps deprecating/removing
+// the old API.
+em_promise_t emscripten_dlopen_promise(const char* filename, int flags) {
+  em_promise_t p = emscripten_promise_create();
+  emscripten_dlopen(filename, flags, p, promise_onsuccess, promise_onerror);
+  return p;
```

Comment:
In principle users may want to use the returned handle even after the promise has been fulfilled or rejected. For example, they may want to use it with `EM_PROMISE_MATCH` in another promise callback or they may want to call `emscripten_promise_then` on it at some point in the future.

As a general design principle, I think it will be worth it to always let the user manage the lifetime of handles returned from our APIs. That will make the future C++ API, which will manage the lifetimes of `em_promise_t` objects using RAII, much safer as well.

I can think of three ways to give the user control over the lifetime of the returned promise here. First, you could return `emscripten_promise_then(p, nop_onsuccess, nop_onerror)`. Second, you could allocate a new promise `ret` and return it after calling `emscripten_promise_resolve(ret, EM_PROMISE_MATCH, p)` to resolve it to match the eventual outcome of `p`. Third, we could add reference counts to our promise handles and we could return `p` directly after increasing its refcount.

I think I prefer the `EM_PROMISE_MATCH` solution because it should be more efficient than the `emscripten_promise_then` solution and does not require complicating the core API with refcounts.

Reply:
Done. PTAL.